libafl/src/corpus/probabilistic_sampling.rs

```diff
@@ -0,0 +1,197 @@
+//! Probabilistic sampling scheduler is a corpus scheduler that feeds the fuzzer
+//! with sampled item from the corpus.
+
+use crate::{
+    bolts::rands::Rand,
+    corpus::{Corpus, CorpusScheduler, FavFactor},
+    inputs::Input,
+    state::{HasCorpus, HasMetadata, HasRand},
+    Error,
+};
+use alloc::string::String;
+use core::marker::PhantomData;
+use hashbrown::HashMap;
+use serde::{Deserialize, Serialize};
+
+/// Conduct reservoir sampling (probabilistic sampling) over all corpus elements.
+#[derive(Debug, Clone)]
+pub struct ProbabilitySamplingCorpusScheduler<I, S, F>
+where
+    I: Input,
+    S: HasCorpus<I> + HasMetadata + HasRand,
+    F: FavFactor<I>,
+{
+    phantom: PhantomData<(I, S, F)>,
+}
+
+/// A state metadata holding a map of probability of corpus elements.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ProbabilityMetadata {
+    /// corpus index -> probability
+    pub map: HashMap<usize, f64>,
+    /// total probability of all items in the map
+    pub total_probability: f64,
+}
+
+crate::impl_serdeany!(ProbabilityMetadata);
+
+impl ProbabilityMetadata {
+    /// Creates a new [`struct@ProbabilityMetadata`]
+    #[must_use]
+    pub fn new() -> Self {
+        Self {
+            map: HashMap::default(),
+            total_probability: 0.0,
+        }
+    }
+}
+
+impl Default for ProbabilityMetadata {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<I, S, F> ProbabilitySamplingCorpusScheduler<I, S, F>
+where
+    I: Input,
+    S: HasCorpus<I> + HasMetadata + HasRand,
+    F: FavFactor<I>,
+{
+    /// Creates a new [`struct@ProbabilitySamplingCorpusScheduler`]
+    #[must_use]
+    pub fn new() -> Self {
+        Self {
+            phantom: PhantomData,
+        }
+    }
+
+    /// Calculate the score and store in `ProbabilityMetadata`
+    #[allow(clippy::cast_precision_loss)]
+    #[allow(clippy::unused_self)]
+    pub fn store_probability(&self, state: &mut S, idx: usize) -> Result<(), Error> {
+        let factor = F::compute(&mut *state.corpus().get(idx)?.borrow_mut())?;
+        if factor == 0 {
+            return Err(Error::DivByZero(
```

Comment:
Should this really be an Error during fuzzing? Like, is there a way for this to happen regularly? If yes, it's probably not a good idea to stop fuzzing,  if not then this can probably be an `Error::IllegalState`

Reply:
Fixed, using `Error::IllegalState` now.